lib/backend/buffer.go

```diff
@@ -280,12 +294,38 @@ func (w *BufferWatcher) Done() <-chan struct{} {
 }
 
 // Close closes the watcher, could
-// be called multiple times
+// be called multiple times, removes the watcher
+// from the buffer queue
 func (w *BufferWatcher) Close() error {
+	w.closeAndRemove(removeAsync)
+	return nil
+}
+
+// closeWatcher closes watcher
+func (w *BufferWatcher) closeWatcher() error {
 	w.cancel()
 	return nil
 }
 
+const (
+	removeSync  = true
+	removeAsync = true
+)
+
+// closeAndRemove closes the watcher, could
+// be called multiple times, removes the watcher
+// from the buffer queue syncronously (used in tests)
+// or asyncronously, used in prod, to avoid potential deadlocks
+func (w *BufferWatcher) closeAndRemove(sync bool) error {
+	w.closeWatcher()
+	if removeSync {
```

Comment:
This will always evaluate to `true` because `removeSync` is a constant defined above. Did you mean `if sync` here?